jenkins/bootstrap.py

```diff
@@ -0,0 +1,438 @@
+        if val and isinstance(val, dict):
+            data['metadata'] = val
+    gsutil.UploadJson(paths.finished, data)
+
+
+
+
+def TestInfra(*paths):
+    """Return path relative to root of test-infra repo."""
+    return os.path.join(ORIG_CWD, os.path.dirname(__file__), '..', *paths)
+
+
+def Node():
+    if 'NODE_NAME' not in os.environ:
+        os.environ['NODE_NAME'] = ''.join(socket.gethostname().split('.')[:1])
+    return os.environ['NODE_NAME']
+
+
+def Version():
+    version_file = 'version'
+    if os.path.isfile(version_file):
+        # TODO(fejta): is this code path ever used?
+        with open(version_file) as fp:
+            return fp.read().strip()
+
+    version_script = 'hack/lib/version.sh'
+    if os.path.isfile(version_script):
+        cmd = [
+            'bash', '-c', (
+"""
+set -o errexit
+set -o nounset
+export KUBE_ROOT=.
+source %s
+kube::version::get_version_vars
+echo $KUBE_GIT_VERSION
+""" % version_script)
+        ]
+        return Subprocess(cmd, output=True).strip()
+
+    return 'unknown'
+
+
+class CIPath(object):
+    base = 'gs://kubernetes-jenkins/logs'
+    build_link = None
+    build_result_cache = None
+
+    def __init__(self, job, build):
+        self.artifacts = os.path.join(self.base, job, build, 'artifacts')
+        self.started = os.path.join(self.base, job, build, 'started.json')
+        self.finished = os.path.join(self.base, job, build, 'finished.json')
+        self.build_log = os.path.join(self.base, job, build, 'build-log.txt')
+        self.latest = os.path.join(self.base, job, 'latest-build.txt')
+        self.result_cache = os.path.join(
+            self.base, job, 'jobResultsCache.json')
+        self.build_latest = self.latest
+
+
+class PRPath(object):
+    base = 'gs://kubernetes-jenkins/pr-logs'
+
+    def __init__(self, job, build, pull):
+        self.build_path = os.path.join(
+            self.base, 'pull', pull, job, build)
+        self.artifacts = os.path.join(self.build_path, 'artifacts')
+        self.started = os.path.join(self.build_path, 'started.json')
+        self.finished = os.path.join(self.build_path, 'finished.json')
+        self.build_log = os.path.join(self.build_path, 'build-log.txt')
+
+        self.latest = os.path.join(
+            self.base, 'directory', job, 'latest-build.txt')
+        self.result_cache = os.path.join(
+            self.base, 'directory', job, 'jobResultsCache.json')
+        self.build_latest = os.path.join(
+            self.base, 'pull', pull, job, 'latest-build.txt')
+        self.build_result_cache = os.path.join(
+            self.base, 'pull', pull, job, 'jobResultsCache.json')
+        self.build_link = os.path.join(
+            self.base, 'directory', job, '%s.txt' % build)
+
+
+def Uniq():
+    """Return a probably unique suffix for the process."""
+    return '%x-%d' % (hash(Node()), os.getpid())
+
+
+BUILD_ENV = 'BUILD_NUMBER'  # TODO(fejta): retire this magic environment variable
+
+def Build(start):
```

Comment:
I'd like to punt on this...

I prefer accessors to be something like `sandwich()` rather than something like `GetCurrentSandwichToConsume()` which normally works great but the confusing bit here is that Build() is both a verb and noun